### PR TITLE
gh-108901: Add `Signature.from_code` method

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-108901

----
:books: Documentation preview :books:: https://cpython-previews--108902.org.readthedocs.build/

### CODE REVIEW

**Comment by sobolevn** [comment]
Failures are not related: https://github.com/python/cpython/issues/108888

File: Lib/inspect.py
Comment (merwok) [inline Lib/inspect.py]: Can you add a stacklevel parameter? (probably `2`)

**Review by vstinner — COMMENTED** [review]
Please restrict this PR to add from_code(). Once it's merged, you can write a PR to deprecate.

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]: I prefer error messages which show the argument type, like:

```
return TypeError('code object expected, got {type(co)}')
```

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]: Please fix the indentation :-)

File: Doc/library/inspect.rst
Comment (vstinner) [inline Doc/library/inspect.rst]: You can suggest using Signature.from_callable() when a function is available.

In from_callable() doc, you may mention from_code() but with a quick reminder about limitations (no default values nor annotations). Or maybe don't mention it there :-)

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]: I don't think that this comment is relevant in the _signature_from_code() function.

If you care, add a comment in the "is not None" code path to explain that it's the _signature_from_function() code path?

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]: Private functions don't have to check their arguments, you already do the exact same check before calling the function. If users use private functions, they are on their own.

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]:
```suggestion
        raise ValueError("no signature found for built-in function {!r}".format(func))
```

File: Lib/inspect.py
Comment (vstinner) [inline Lib/inspect.py]: You may rename it to _signature_from_builtin_func() ;-)

File: Lib/test/test_inspect.py
Comment (vstinner) [inline Lib/test/test_inspect.py]:
```suggestion
    def test_signature_from_code(self):
```

File: Lib/test/test_inspect.py
Comment (vstinner) [inline Lib/test/test_inspect.py]: Just move this test in the previous function, i don't think that it's worth it to separate it.

**Review by vstinner — COMMENTED** [review]
I like your overall change, here is a first review.

File: Doc/library/inspect.rst
Comment (AlexWaygood) [inline Doc/library/inspect.rst]: Typo:

```suggestion
       they will always be omitted from the signature.
```

File: Lib/inspect.py
Comment (sobolevn) [reply Lib/inspect.py]: This is not actually something I change :)

This is git acting strange, because I just moved some code above. 

I would prefer not to touch these parts. They are unrelated. I try to move code again to change the diff.

**Comment by sobolevn** [comment]
I move my code, so there should be no unrelated diffs now :)

File: Doc/library/inspect.rst
Comment (vstinner) [inline Doc/library/inspect.rst]: Thanks, it's more explicit and propose a concrete action to avoid these limitations!

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer a second core dev review, since I don't know well this module.

**Comment by vstinner** [comment]
@pablogsal @AlexWaygood @1st1: Would you be interested to review this change?

File: Doc/library/inspect.rst
Comment (AlexWaygood) [inline Doc/library/inspect.rst]: Use italics rather than backticks when referring to parameters

```suggestion
       for a given :class:`code object <types.CodeType>` *co*.
```

File: Doc/library/inspect.rst
Comment (AlexWaygood) [inline Doc/library/inspect.rst]: Lead with the important thing you're trying to tell the reader, then follow up with the explanation of why it's the case:

```suggestion
       Default values and annotations
       will not be included in the signature,
       since code objects have no knowledge of such things.
```

File: Doc/library/inspect.rst
Comment (AlexWaygood) [inline Doc/library/inspect.rst]: grammar fix:

```suggestion
       when a function object is available,
       as ``Signature.from_callable()`` does not have the same limitations.
```

File: Lib/inspect.py
Comment (AlexWaygood) [inline Lib/inspect.py]: Weird indentation:

```suggestion
    return _signature_from_code(cls,
                                func.__code__,
                                globals=globals,
                                locals=locals,
                                eval_str=eval_str,
                                is_duck_function=is_duck_function,
                                func=func)
```

File: Lib/inspect.py
Comment (AlexWaygood) [inline Lib/inspect.py]: PEP-257 nit:

```suggestion
    """Private helper function for constructing Signatures from code objects."""
```

File: Lib/inspect.py
Comment (AlexWaygood) [inline Lib/inspect.py:2452]: I prefer to use affirmative conditions where possible (`if func is None; else` rather than `if func is not None; else`) -- the double negatives make my brain go a little fuzzy otherwise 😄

```suggestion
    if func is None:
        annotations = {}
        defaults = None
        kwdefaults = None
    else:
        annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
        defaults = func.__defaults__
        kwdefaults = func.__kwdefaults__
```

File: Lib/inspect.py
Comment (AlexWaygood) [inline Lib/inspect.py:3132]: (No change requested)

Generally we prefer the imperative mood for docstrings, but this is in keeping with the style of the other docstrings in this class, so I think it works well 👍

File: Lib/inspect.py
Comment (AlexWaygood) [inline Lib/inspect.py]: The usual style for Python `TypeError`s is to quote the _name_ of the incorrect type rather than the _repr_ of the incorrect type:

```pycon
>>> 'foo' + 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate str (not "int") to str
```

(also, we usually use "was" rather than "is")

```suggestion
            raise TypeError(f'code object was expected, got {type(co).__name__!r}')
```

**Review by AlexWaygood — COMMENTED** [review]
A few nits:

File: Lib/test/test_inspect.py
Comment (AlexWaygood) [inline Lib/test/test_inspect.py:3699]: Well, I worry that the reason this wasn't implemented in the original PEP-362 implementation is because the signatures produced for `func1`, `func2` and `func5` here are all incorrect, due to the fact that certain parameters have default values but this isn't reflected in the produced signatures :/

`inspect.getargs()` is [used a fair bit in third-party projects](https://grep.app/search?q=inspect.getargs&words=true&filter[lang][0]=Python). In the vast majority of these uses, they already have access to the function object, so they should probably just use `Signature.from_callable()` instead of this new method. But there are a few uses where they don't have access to the function object -- so I agree that something like this new method _is_ probably needed, as a replacement for those uses:
- https://github.com/0compute/xtraceback/blob/5f4ae11cf21e6eea830d79aed66d3cd91bd013cd/xtraceback/xtracebackframe.py#L53-L55
- https://github.com/ponyorm/pony/blob/6f7620c9e66a4b8d4f7f7ba4a456b733aab9f374/pony/orm/decompiling.py#L793-L798
- https://github.com/parrt/lolviz/blob/301779d39934937795b71ad5d9d62d8a45f7bee1/lolviz.py#L297-L303

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: > the signatures produced for `func1`, `func2` and `func5` here are all incorrect

Hmmm... well, they are actually valid signatures _for the code object_, since if you pass the code object to `exec()`, it doesn't take any notice of whether the original function had a default value or not:

```pycon
>>> def foo(x=42): print(x)
...
>>> foo()
42
>>> c = foo.__code__
>>> exec(c)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() missing 1 required positional argument: 'x'
```

Why anybody would care what the signature of the code object itself is, though, I'm not sure -- ISTM that everybody trying to get the signature from a code object is just using the code object as the next-best-thing for the function the code object belongs to. I don't think many people are passing code objects directly to `exec()` It's not even really possible to pass a parameter to a code object you're calling via `exec()` AFAIK.

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: Maybe we just need to flag more forcefully in the docs the limitations of this new constructor method. I can offer some suggestions once you've addressed the review comments I've already posted 👍

File: Lib/inspect.py
Comment (merwok) [reply Lib/inspect.py:2452]: But `if param is not None` is extremely common, and the negated negative makes it a positive!
It’s fine if the code wants to address first the case where param is passed something (not None).

File: Lib/inspect.py
Comment (merwok) [reply Lib/inspect.py:3132]: The imperative meaning:
```suggestion
        """Construct Signature for the given code object.
```

Comment (AlexWaygood) [reply Lib/inspect.py:3132]: > The imperative meaning

Indeed, but, as I say, all other constructor methods on this class currently use "Constructs" and "Creates" rather than "Construct" and "Create" in their docstrings:

https://github.com/python/cpython/blob/1fb20d42c58924e2e941622b3539645c7b843e0e/Lib/inspect.py#L3052

https://github.com/python/cpython/blob/1fb20d42c58924e2e941622b3539645c7b843e0e/Lib/inspect.py#L3105

https://github.com/python/cpython/blob/1fb20d42c58924e2e941622b3539645c7b843e0e/Lib/inspect.py#L3119

I think consistency within the class is more important here, so while I prefer the imperative mood in general, in this specific instance I prefer the wording that @sobolevn already has.

File: Lib/inspect.py
Comment (AlexWaygood) [reply Lib/inspect.py:2452]: I still prefer my suggestion, but I don't feel strongly on this one. We can let @sobolevn decide which he prefers.

Comment (sobolevn) [reply Lib/inspect.py:2452]: Yeap, I prefer to keep this as-is, because `func` is `not None` most of the time.

File: Lib/inspect.py
Comment (sobolevn) [reply Lib/inspect.py:3132]: I used the same style as in `from_callable`

File: Lib/test/test_inspect.py
Comment (sobolevn) [reply Lib/test/test_inspect.py:3699]: I added a `.. note::` notation to highlight it even better.

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: The more I think about this, the more concerned I am :/

The existing `getargs()` function that we have in `inspect` gives _accurate_ information. It doesn't _claim_ to be able to tell you everything about the signature: it only _claims_ to be able to tell you _some_ information about the signature:
- It gives you a list of argument names
- It tells you the name of the `*vararg` argument, if there is one
- It tells you the name of the `**kwarg` argument, if there is one

This new function, however, _claims_ to be able to give you a complete and accurate signature from a code object (in that it constructs a `Signature` object from a code object). That's impossible, since code objects don't have information about which arguments have default values, and so the function gives you _inaccurate information_.

I'm not sure having a prominent note in the docs is sufficient to fix this issue: adding a new constructor that we know gives inaccurate results seems very troubling to me. Perhaps a better solution would be to document `getargs` and promote it to public API?

Comment (sobolevn) [reply Lib/test/test_inspect.py:3699]: `getargs` is fundamentally broken:

```python
>>> import inspect
>>> def a(x, /, *, y): ...

>>> inspect.getargs(a.__code__)
Arguments(args=['x', 'y'], varargs=None, varkw=None)
```

It does not give you *any* helpful information about `x` and `y`.
So, promoting it to a signature is much better:

```python
>>> inspect.Signature.from_code(a.__code__)
<Signature (x, /, *, y)>
```

You can actually know the parameters' kinds. Now, you can actually call it:

```python
>>> import types
>>> types.FunctionType(a.__code__, {}, 'a')(1, y=2)
```

Compare it with the failed attempt, when you *don't* know the parameter kinds:

```python
>>> types.FunctionType(a.__code__, {}, 'a')(x=1, y=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a() got some positional-only arguments passed as keyword arguments: 'x'
```

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: True. This new API would provide helpful information that is not currently offered by `getargs`.

However. Does it make sense to think of code objects even having a "signature"? _Functions_ have signatures, and you can _construct_ functions from code objects. But as your snippet above demonstrates, you can't directly call code objects (except via `exec()`, but nobody does that really) -- you have to create a new function from the code object in order to get something that can be called.

Which "signature" are we trying to get here? The signature of the function from which the code object originally came (if it came from a code object), or the signature a new function object would have if we created one from the code object? Getting a "signature from a code object" feels like quite a confusing concept to me.

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: Another option: we could document that if people want to get "signature information" from a code object `co`, they just need to do this:

```py
inspect.signature(types.FunctionType(co, {}))
```

That way, they are forced to be explicit that code objects don't really have a "signature": you need to construct a new function from the code object in order to get a code object that actually has a signature

Comment (sobolevn) [reply Lib/test/test_inspect.py:3699]:
```python
>>> import types, inspect
>>> def a(x: int, /, *, y: str = '') -> None: ...
... 
>>> inspect.signature(types.FunctionType(a.__code__, {}))
<Signature (x, /, *, y)>
```

It is up to you to decide whether or not we should do it. I consider this PR done from my side.

Comment (AlexWaygood) [reply Lib/test/test_inspect.py:3699]: Currently, I'm leaning towards not adding this alternative constructor, and just documenting that people should call `inspect.signature(types.FunctionType(co, {}))` if they need to get signature information from a code object `co`, and the function object is not available for whatever reason.

One of the third-party projects that uses `getargs` is Quora: https://github.com/quora/qcore/blob/e12a4929c8e75ec2dd878c06270cedcac72a643e/qcore/inspection.py#L130-L148. I'd be curious to know if @JelleZijlstra has any thoughts on whether `getargs` should be documented or deprecated and, if the latter, whether we should add this `Signature.from_code()` method or just add documentation along the lines of my sugestion.

Comment (JelleZijlstra) [reply Lib/test/test_inspect.py:3699]: Any code using that qcore function should really be migrated to use `inspect.signature` directly. As the docstring notes, the function doesn't deal correctly with kwonly arguments.

I think we don't need `Signature.from_code`. My experience is that it is rare that you have *only* a code object without a function it came from, and code that manipulates code objects directly should generally be migrated to a more modern inspect API.

**Comment by AlexWaygood** [comment]
Thanks for the PR, @sobolevn -- it was well done. But I think adding a constructor that implies you can construct a complete `Signature` from a code object is too misleading. Let's go with documenting that you should call `inspect.signature(types.FunctionType(co, {}))` in the rare cases that you don't have access to the original function object.

**Pull request closed without merging**
